pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainGist failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://mesolite.cluster.engflow.com/invocations/default/c3f8d6d7-7034-4afe-a084-87fb4adee51e?testReportRun=10&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [4e7fa9411f645a8bf0f8baf0bb1147118d0cf730](https://github.com/cockroachdb/cockroach/commits/4e7fa9411f645a8bf0f8baf0bb1147118d0cf730):


```
        										AS tab_1333 (col_2314, col_2315)
        							)
        								AS tab_1334 (col_2318, col_2319)
        					),
        				with_289 (col_2321)
        					AS (
        						SELECT
        							*
        						FROM
        							(
        								VALUES
        									('null':::JSONB),
        									(
        										'[{"baz": true}, {"IDbns": "20svi", "bar": false}, [false], 1.3592211400773624]':::JSONB
        									)
        							)
        								AS tab_1335 (col_2321)
        					)
        			SELECT
        				NULL AS col_2324
        			FROM
        				(
        					VALUES
        						(ARRAY[1867091974:::OID], '97AF088F/1811AC77':::PG_LSN),
        						(
        							ARRAY[
        								2214969618:::OID,
        								1428061439:::OID,
        								982559999:::OID,
        								438641179:::OID,
        								1854142403:::OID,
        								3760194544:::OID,
        								1259832267:::OID,
        								2335079949:::OID
        							],
        							'66529C23/C3210E61':::PG_LSN
        						)
        				)
        					AS tab_1336 (col_2322, col_2323)
        			ORDER BY
        				tab_1336.col_2323 NULLS LAST, tab_1336.col_2322 NULLS FIRST
        			LIMIT
        				48:::INT8
        				INTO p3;
        			p2 :=
        				'[-1.5096147,2.3610923,-0.15536839,0.81638354,-0.5368428,1.5952786,0.611992,0.074733615,-1.8734491,0.07080379,1.2771158,-0.65509015,-0.06289779,-0.07033529,-1.9963577,-1.0379329,1.9180187,-1.4299632,-0.021407776,-0.5376049,-0.018661128,-0.4059227,0.07620788,-0.07412413,1.5542547,-0.22970118,0.92355376,0.30682942,0.23024923,0.273495,-1.2326783,-0.102891125,-1.272455,0.9886574,-1.2267556,0.101549424,2.2493749,1.7524126,-0.118606105,0.60557747,0.7269001,-0.98627853,-0.39184454,1.7703243,1.3044024,0.97655386,-1.0372893,0.46263373,1.4073232,-0.59994483,-0.93902314,1.2216765,-0.59301805,-1.1546152,-0.3970478,0.089980766,0.6585908,-1.8406031,-1.0524336,1.6151738,0.39692327,1.6177765,0.06626565,-1.7498548,-0.24021733,-1.1281501,2.377423,0.09259317,1.7859917,-0.08931055,0.6882081,0.5805557,1.63982,-2.6345205,-0.5080324,-1.0613208,-0.87378806,0.8199359,-0.7563248,0.11358416,-0.3117237,-0.5690553,-0.008386267,-2.3962421,0.815359,0.33317405,0.24290873,-1.7459618,-0.94762594,1.1929547,-1.8035455,-1.7738584,0.32999498,1.7040949,-1.4326301,-0.74527293,0.026974726,0.070806146,-0.69830686,0.28819484,-2.2269404,-0.83064634,1.8583997,-0.15846987,-0.0011759204,-0.108850576,0.50280446,-0.3699565,-1.9908551,0.47601378,-0.007828862,0.75613,-0.49291757,-1.4847175,-0.608364,-0.9667487,-0.91824865,-0.6271924,1.1466898,-0.3529516,0.61695045,-0.4440907,0.27084503,0.8642988,-0.33992964,0.7873189,1.0688857,2.6976647,-2.3155656,-0.8197325,0.7356713,-0.1976852,1.6059363,-0.15287533,-1.5282354,-2.005071,1.2512354,-0.77123475,-0.07280366,0.68306357,-0.56295615,-0.4130632,-1.0921495,0.61181235,0.6606916,-1.0072867,-1.8100282,0.12809765,0.37368065,-0.22214639,0.1124679,0.7314429,1.2227658,0.22288369,-0.06995237,-0.38109228,-0.23337974,1.4615217,-0.19289902,-0.12202455,-0.48978496,-2.1435435,0.7628837,-0.4621557,-0.8147214,0.43568906,-0.40027896,-1.3402889,0.2226399,-0.94099283,-1.2136226,-0.021963406,-1.1120788,-0.0895291,0.22402635,0.8822826,-0.383094,-1.4346329,-0.29248303,-1.082232,2.4637096,0.4698489,0.30381358,-0.3937775,1.6584603,-0.12815559,-0.33583233,-1.5691078,-0.15821847,1.9303252,3.3542569,0.10513917,1.7462816,-0.16112974,1.0982801,1.2260132,-1.3247608,-1.1888053,1.3799189,0.8517196,-0.22808391,0.4353898,0.3292041]':::VECTOR;
        			p0 := p0;
        		END;
        	$$;
--- FAIL: TestExplainGist/main (7.13s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-64200

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21358681?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21358681?buildTab=artifacts#/) on master @ [b852382ca5d1b265a84370597f0bd77c0ab110cb](https://github.com/cockroachdb/cockroach/commits/b852382ca5d1b265a84370597f0bd77c0ab110cb):


```
        											)::LTREE
        										)
        								)
        									AS tab_1121 (col_1873)
        						)
        				SELECT
        					'':::REFCURSOR AS col_1874,
        					p1 AS col_1875,
        					B'001110010000011011000000010001000010' AS col_1876,
        					p6 AS col_1877,
        					decl_514 AS col_1878,
        					p6 AS col_1879,
        					(-4865):::INT8 AS col_1880,
        					p3 AS col_1881,
        					'08:08:11.36788+02:46':::TIMETZ AS col_1882,
        					'good day':::greeting AS col_1883,
        					p0 AS col_1884
        				FROM
        					with_253 AS cte_ref_69
        				WHERE
        					false
        				ORDER BY
        					cte_ref_69.col_1871 DESC, cte_ref_69.col_1871 NULLS FIRST, cte_ref_69.col_1872
        				LIMIT
        					28:::INT8;
        				NULL;
        			ELSE
        				SELECT
        					st_asmvt(e'\\x59c54e2848a940':::BYTES, e'\\x0b107f':::BYTES::STRING, NULL::INT8)::BYTES AS col_1885,
        					'good day':::greeting AS col_1886
        				FROM
        					defaultdb.public.seed AS tab_1122
        				WHERE
        					tab_1122._bool
        				GROUP BY
        					tab_1122._bytes, tab_1122._enum
        				LIMIT
        					20:::INT8;
        			END IF;
        			SELECT
        				(-1):::DECIMAL AS col_1887,
        				tab_1123._float8 AS col_1888,
        				e'K\x14h\x0e':::REFCURSOR AS col_1889,
        				p4 AS col_1890,
        				p3 AS col_1891
        			FROM
        				defaultdb.public.seed@seed__int8__float8__date_idx AS tab_1123;
        		END;
        	$funcbody$;
--- FAIL: TestExplainGist/main (4.83s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/logictest/testdata/logic_test/udf_regressions
+++ b/pkg/sql/logictest/testdata/logic_test/udf_regressions
@@ -198,7 +198,6 @@ SELECT count(descriptor) FROM system.descriptor WHERE id = $dropped_fn_id;
 ----
 0
 
-
 subtest end
 
 subtest regression_94146
@@ -453,7 +452,6 @@ SELECT f_103869('sq_103869')
 ----
 2
 
-
 subtest end
 
 # Regression test for #104927. Correctly resolve table references in UDFs as
@@ -488,7 +486,6 @@ query T
 SELECT f104927()
 ----
 [{"i": 1, "s": "foo"}]
-
 
 subtest end
 
@@ -543,7 +540,6 @@ $$
 
 subtest end
 
-
 # Regression tests for #105259 and #107654. Do not type-check subqueries in UDFs
 # outside optbuilder. Doing so can cause internal errors.
 subtest regression_105259
@@ -566,7 +562,6 @@ CREATE FUNCTION f() RETURNS VOID LANGUAGE SQL AS $$
 $$
 
 subtest end
-
 
 # Regression test for #108297. UDFs with VOID return types should succeed when
 # the last statement returns columns of any type.
@@ -864,5 +859,54 @@ DROP FUNCTION sc_62227.next_id;
 DROP FUNCTION sc_62227.next_id_sql;
 DROP SEQUENCE sc_62227.myseq;
 DROP SCHEMA sc_62227;
+
+subtest end
+
+subtest regression_170852
+
+# Regression test for #170852. Building a routine whose body contains a
+# NULL-valued REGCLASS expression (e.g. NULL::REGCLASS, or an uninitialized
+# OUT parameter of type REGCLASS) should not fail with an internal error.
+# maybeTrackRegclassDependenciesForViews evaluates constant REGCLASS
+# expressions and previously asserted the result was a *tree.DOid, but a NULL
+# REGCLASS evaluates to tree.DNull.
+statement ok
+CREATE FUNCTION f_170852_null() RETURNS INT LANGUAGE plpgsql AS $$
+DECLARE
+  v INT;
+BEGIN
+  SELECT nextval(NULL::REGCLASS) INTO v;
+  RETURN v;
+END;
+$$;
+
+# An uninitialized OUT parameter of type REGCLASS also evaluates to NULL when
+# the routine body is built.
+statement ok
+CREATE FUNCTION f_170852_out(OUT p0 REGCLASS, IN p1 BOOL) RETURNS REGCLASS
+LANGUAGE plpgsql AS $$
+BEGIN
+  p0 := p0;
+END;
+$$;
+
+# The same NULL handling applies to SQL UDFs.
+statement ok
+CREATE FUNCTION f_170852_sql() RETURNS REGCLASS LANGUAGE sql AS $$
+  SELECT NULL::REGCLASS;
+$$;
+
+# Empty function body should work too.
+statement ok
+CREATE FUNCTION f_170852_empty_body(OUT p0 REGCLASS, IN p1 BOOL) RETURNS REGCLASS
+LANGUAGE plpgsql AS $$
+BEGIN
+END;
+$$;
+
+statement ok
+DROP FUNCTION f_170852_null;
+DROP FUNCTION f_170852_out;
+DROP FUNCTION f_170852_sql;
 
 subtest end

--- a/pkg/sql/opt/optbuilder/builder.go
+++ b/pkg/sql/opt/optbuilder/builder.go
@@ -588,7 +588,13 @@ func (b *Builder) maybeTrackRegclassDependenciesForViews(texpr tree.TypedExpr) {
 	if err != nil {
 		panic(err)
 	}
-	// eval.Expr on a REGCLASS expression returns a *tree.DOid.
+	// A NULL REGCLASS references no object, so there is no dependency to track.
+	// This happens, for example, with a NULL::REGCLASS literal or an
+	// uninitialized OUT parameter of type REGCLASS in a routine body.
+	if regclass == tree.DNull {
+		return
+	}
+	// eval.Expr on a non-NULL REGCLASS expression returns a *tree.DOid.
 	// Use the OID directly for resolution rather than DOid.String(), which
 	// returns only the object name and drops the schema prefix (e.g. returns
 	// "myseq" instead of "sc.myseq"), causing unqualified lookups to fail

